File: fast_sync_dart_client/fast_sync_client/lib/src/absraction/models/sync_payload.dart
import 'package:fast_sync_client/fast_sync_client.dart';

class SyncPayload<T extends ISyncableObject> {
  Map<String, List<T>> _data;
  SyncOperationMetadata operationMetadata;
  bool hasData = false;

  SyncPayload(
      {Map<String, List<T>>? data, SyncOperationMetadata? operationMetadata})
      : _data = data ?? {},
        operationMetadata = operationMetadata ?? SyncOperationMetadata();

  void pushObjects(String type, List<T> entities) {
    if (entities.isNotEmpty) {
      Map<String, dynamic> typeMetadataJson = {
        "syncZone": FastSync.getTypeSyncZone(type),
        "type": type,
        'id': "id",
        'version': 999,
        'timestamp': 999,
        'syncOperation': 999,
      };
      operationMetadata.setMetadata(type, buildMetadata(typeMetadataJson));
      _data[type] = entities;
      hasData = true;
    }
  }

  List<T> getObjectsForType(String type) {
    return _data.containsKey(type) ? _data[type]! : [];
  }

  List<String> getSyncedTypes() {
    return _data.keys.toList();
  }

  ISyncMetadata getTypeMetadata(String type) {
    final metadata = operationMetadata.getTypeMetadata(type);
    return metadata;
  }

  factory SyncPayload.fromJson(Map<String, dynamic> json) {
    SyncConfiguration configuration = FastSync.getSyncConfiguration();
    final dataMap = json['data'] as Map<String, dynamic>;
    final operationMetadata =
        SyncOperationMetadata.fromJson(json['operationMetadata']);

    final data = <String, List<T>>{};
    dataMap.forEach((key, value) {
      final list = (value as List).cast<Map<String, dynamic>>();
      Function typeFromJson = configuration.getTypeForFromJsonFunction(key);
      List<T> items = List<T>.from(list.map((item) => typeFromJson.call(item)));
      data[key] = items;
    });

    return SyncPayload(
      data: data,
      operationMetadata: operationMetadata,
    );
  }

  Map<String, dynamic> toJson() {
    SyncConfiguration configuration = FastSync.getSyncConfiguration();
    Map<String, List<Map<String, dynamic>>> jsonDataMap = {};

    _data.forEach((key, value) {
      Function typeToJson = configuration.getTypeForToJsonFunction(key);
      List<Map<String, dynamic>> itemsTypeJson = value
          .map((item) => typeToJson.call(item) as Map<String, dynamic>)
          .toList();
      jsonDataMap[key] = itemsTypeJson;
    });
    return {
      'operationMetadata': this.operationMetadata.toJson(),
      'data': jsonDataMap,
    };
  }
}

File: fast-sync-typescript-client-sdk/packages/fast-sync-client/src/abstraction/models/sync_payload.ts
import { FastSync } from "../../implementation/fast_sync";
import { SyncMetadata } from "../../implementation/metadata/syncable_metadata";
import { isNullOrUndefined } from "../../implementation/utils/utils";
import { ISyncMetadata } from "../metadata/isync_metadata";
import { ISyncableObject } from "../metadata/isyncable_object";
import { SyncOperationMetadata } from "./sync_operation_metadata";

export class SyncPayload {
  public data: Map<string, any[]>;
  operationMetadata: SyncOperationMetadata;
  hasData: boolean = false;

  static create(syncPayload: SyncPayload) {
    let payload = new SyncPayload();
    payload.data = syncPayload.data;
    payload.operationMetadata = SyncOperationMetadata.create(
      syncPayload.operationMetadata
    );
    return payload;
  }

  constructor(
    data?: Map<string, any[]>,
    operationMetadata?: SyncOperationMetadata
  ) {
    this.data = data ?? new Map<string, any[]>();
    this.operationMetadata = operationMetadata ?? new SyncOperationMetadata();
  }

  pushObjects<T extends ISyncableObject>(type: string, entities: T[]): void {
    if (entities.length > 0) {
      let syncMetadata = new SyncMetadata(
        type,
        999,
        FastSync.getTypeSyncZone(type)
      );
      this.operationMetadata.setMetadata(
        type,
        this.buildMetadata(syncMetadata)
      );
      this.data.set(type, entities);
      this.hasData = true;
    }
  }

  getObjectsForType(type: string): any[] {
    if (this.data.hasOwnProperty(type) && !isNullOrUndefined(this.data[type])) {
      return this.data[type];
    } else {
      return [];
    }
  }

  getSyncedTypes(): string[] {
    return Object.keys(this.data);
  }

  getTypeMetadata(type: string): ISyncMetadata {
    return this.operationMetadata.getTypeMetadata(type);
  }

  private buildMetadata(metadataJson: { [key: string]: any }): ISyncMetadata {
    // Assuming the implementation of buildMetadata is defined elsewhere in your codebase
    // Replace this with the actual implementation
    return metadataJson as ISyncMetadata;
  }
}
